retrieval/state.go

```diff
@@ -0,0 +1,220 @@
+	case PhaseDataRetrieval:
+		return "data retrieval"
+	default:
+		return "unknown"
+	}
+}
+
+// RetrievalState tracks diagnostic information about IPFS content retrieval operations.
+// It is safe for concurrent use and maintains monotonic stage progression.
+// Use ContextWithState to add tracking to a context, and StateFromContext
+// to retrieve the state for updates or inspection
+type RetrievalState struct {
+	// ProvidersFound tracks the number of providers discovered for the content.
+	ProvidersFound atomic.Int32
+	// ProvidersAttempted tracks the number of providers we tried to connect to.
+	ProvidersAttempted atomic.Int32
+	// ProvidersConnected tracks the number of providers successfully connected.
+	ProvidersConnected atomic.Int32
+
+	// phase tracks the current retrieval phase (stored as int32)
+	phase atomic.Int32
+
+	// mu protects failedProviders slice during concurrent access
+	mu sync.RWMutex
+	// Track specific provider failures (limited to first few for brevity)
+	failedProviders []peer.ID
+}
+
+// NewRetrievalState creates a new RetrievalState initialized to PhaseInitializing.
+// The returned state is safe for concurrent use.
+func NewRetrievalState() *RetrievalState {
+	rs := &RetrievalState{
+		failedProviders: make([]peer.ID, 0, MaxFailedProvidersToTrack),
+	}
+	rs.phase.Store(int32(PhaseInitializing))
+	return rs
+}
+
+// SetPhase updates the current retrieval phase to the given phase.
+// The phase progression is monotonic - phases can only move forward, never backward.
+// If the provided phase is less than or equal to the current phase, this is a no-op.
+// This method is safe for concurrent use.
+func (rs *RetrievalState) SetPhase(phase RetrievalPhase) {
+	newPhase := int32(phase)
+	for {
+		current := rs.phase.Load()
+		// Only update if the new phase is greater (moving forward)
+		if newPhase <= current {
+			return
+		}
+		// Try to update atomically
+		if rs.phase.CompareAndSwap(current, newPhase) {
+			return
+		}
+		// If CAS failed, another goroutine updated it, loop will check again
+	}
+}
+
+// GetPhase returns the current retrieval phase.
+// This method is safe for concurrent use.
+func (rs *RetrievalState) GetPhase() RetrievalPhase {
+	return RetrievalPhase(rs.phase.Load())
+}
+
+// AddFailedProvider records a provider peer ID that failed to deliver the requested content.
+// Only the first MaxFailedProvidersToTrack providers are kept to prevent unbounded memory growth.
+// This method is safe for concurrent use.
+func (rs *RetrievalState) AddFailedProvider(peerID peer.ID) {
+	rs.mu.Lock()
+	defer rs.mu.Unlock()
+	if len(rs.failedProviders) < MaxFailedProvidersToTrack {
+		rs.failedProviders = append(rs.failedProviders, peerID)
+	}
+}
+
+// GetFailedProviders returns the list of failed providers
+func (rs *RetrievalState) GetFailedProviders() []peer.ID {
+	rs.mu.RLock()
+	defer rs.mu.RUnlock()
+	result := make([]peer.ID, len(rs.failedProviders))
+	copy(result, rs.failedProviders)
+	return result
```

Comment:
`slices.Clone` is generally preferred to the `make` + `copy` pattern. Efficiency/performance is same.
```suggestion
	return slices.Clone(rs.failedProviders)
```